python/ray/llm/_internal/batch/stages/vllm_engine_stage.py

```diff
@@ -0,0 +1,530 @@
+    StatefulStage,
+    StatefulStageUDF,
+)
+from ray.util.scheduling_strategies import PlacementGroupSchedulingStrategy
+
+if TYPE_CHECKING:
+    from PIL import Image
+
+logger = logging.getLogger(__name__)
+
+
+class vLLMEngineWrapper:
+    """Wrapper around the vLLM engine to handle async requests.
+
+    Args:
+        *args: The positional arguments for the engine.
+        max_pending_requests: The maximum number of pending requests in the queue.
+        runtime_env: The runtime environment to use for the vLLM engine.
+        **kwargs: The keyword arguments for the engine.
+    """
+
+    @dataclass(frozen=True)
+    class LLMRequest:
+        """A request to the LLM wrapper."""
+
+        # The request ID for the LLM engine (unique per replica).
+        request_id: int
+        # The index of the request in the batch.
+        idx_in_batch: int
+        # The full prompt string (with chat template applied if any).
+        prompt: str
+        # The images inputs for the multimodal model.
+        images: List["Image.Image"]
+        # The tokenized prompt IDs. If None, then the string prompt will be
+        # tokenized by the LLM engine. This is not recommended for performance reasons.
+        prompt_token_ids: Optional[List[int]]
+        # The sampling or pooling parameters. Use Any to avoid importing vLLM.
+        params: Any
+
+    def __init__(
+        self,
+        *args,
+        idx_in_batch_column: str,
+        max_pending_requests: int = -1,
+        runtime_env: Optional[Dict[str, Any]] = None,
+        **kwargs,
+    ):
+        self.request_id = 0
+        self.idx_in_batch_column = idx_in_batch_column
+        self.task_type = kwargs.get("task", "generate")
+
+        # Setup os.environ before importing vLLM to make sure the environment
+        # variables are effective
+        if runtime_env is not None and "env_vars" in runtime_env:
+            os.environ.update(runtime_env["env_vars"])
+
+        # Lazy import vLLM here.
+        self.vllm = importlib.import_module("vllm")
+
+        # Construct PoolerConfig if override_pooler_config is specified.
+        if self.task_type == "embed" and "override_pooler_config" in kwargs:
+            kwargs["override_pooler_config"] = self.vllm.config.PoolerConfig(
+                **kwargs["override_pooler_config"]
+            )
+
+        # Initialize the vLLM engine.
+        engine_args = self.vllm.AsyncEngineArgs(
+            *args,
+            **kwargs,
+            disable_log_requests=True,
+        )
+        self.engine = self.vllm.AsyncLLMEngine.from_engine_args(engine_args)
+
+        # Determine the generate function based on vLLM v0 or v1.
+        if self.vllm.envs.VLLM_USE_V1:
+            self._generate_async = self.generate_async_v1
+        else:
+            self._generate_async = self.generate_async_v0
+
+        # The performance gets really bad if there are too many requests in the pending queue.
```

Comment:
why don't we use a semaphore here?

Reply:
Intuitively Queue should have better performance than semaphore especially when the size large I guess, but not verified.

Reply:
discussed offline. Let's explore using semaphore here. Seems a more elegant solution.